src/iminuit/cost.py

```diff
@@ -567,11 +572,23 @@ def _prepare_data(self, value):
 
         return n
 
+    def _pred(self, args):
+        X = np.meshgrid(*self._xe) if self._ndim > 1 else self._xe
```

Comment:
Is recreating `X` here for every iteration expensive?  Maybe create it once in `__init__`?

Reply:
Yes, it is, good point. I also found a few other places where the code is not caching results properly that do not vary a lot. The cost functions need a redesign in regard to performance. The class hierarchy is also leaking abstractions everywhere.